controllers/scenario/scenario_adapter.go

```diff
@@ -0,0 +1,146 @@
+)
+
+// Adapter holds the objects needed to reconcile a Release.
+type Adapter struct {
+	application *applicationapiv1alpha1.Application
+	scenario    *v1alpha1.IntegrationTestScenario
+	logger      logr.Logger
+	client      client.Client
+	context     context.Context
+}
+
+// NewAdapter creates and returns an Adapter instance.
+func NewAdapter(application *applicationapiv1alpha1.Application, scenario *v1alpha1.IntegrationTestScenario, logger logr.Logger, client client.Client,
+	context context.Context) *Adapter {
+	return &Adapter{
+		application: application,
+		scenario:    scenario,
+		logger:      logger,
+		client:      client,
+		context:     context,
+	}
+}
+
+// EnsureCreatedScenarioIsValid is an operation that ensures created IntegrationTestScenario is valid
+// in case it is, set its owner reference
+func (a *Adapter) EnsureCreatedScenarioIsValid() (reconciler.OperationResult, error) {
+	if meta.IsStatusConditionTrue(a.scenario.Status.Conditions, gitops.HACBSIntegrationStatusCondition) {
+		a.logger.Info("IntegrationTestScenario is Invalid.",
+			"Scenario", a.scenario)
+		return reconciler.ContinueProcessing()
+	} else {
+		patch := client.MergeFrom(a.scenario.DeepCopy())
+		err := ctrl.SetControllerReference(a.application, a.scenario, a.client.Scheme())
+		if err != nil {
+			a.logger.Error(err, "Error setting owner reference.",
+				"Application.Name", a.application.Name,
+				"Application.Namespace", a.application.Namespace)
+		}
+		err = a.client.Patch(a.context, a.scenario, patch)
+		if err != nil {
+			a.logger.Error(err, "Failed to update Global Candidate for the Scenario",
+				"Scenario.Name", a.scenario.Name)
+			return reconciler.RequeueWithError(err)
+		}
+		return reconciler.ContinueProcessing()
+	}
+
+}
+
+// EnsureGlobalScenarioUpdated is an operation that ensure the IntegrationTestScenario in the Global Candidate List
+// being updated when the IntegrationTestScenario(ITS) gets detected, set owner reference only for valid ITS
+func (a *Adapter) EnsureGlobalScenarioUpdated() (reconciler.OperationResult, error) {
+
+	if a.scenario != nil && meta.IsStatusConditionFalse(a.scenario.Status.Conditions, gitops.HACBSIntegrationStatusCondition) {
+		patch := client.MergeFrom(a.scenario.DeepCopy())
+		for _, refs := range a.scenario.OwnerReferences {
+			a.scenario.OwnerReferences = append(a.scenario.OwnerReferences, refs)
+			err := a.client.Patch(a.context, a.scenario, patch)
+			if err != nil {
+				a.logger.Error(err, "Failed to update Global Candidate for the Scenario",
+					"Scenario.Name", a.scenario.Name)
+				return reconciler.RequeueWithError(err)
+			}
+		}
+
+	}
+
+	return reconciler.ContinueProcessing()
+}
+
+// EnsureEnvironmentExists checks if ITS has environment object defined
+// in case it does > makes sure it exists within the same namespace as ITS
+// otherwise inform about opposite fact.
+func (a *Adapter) EnsureEnvironmentExists() (reconciler.OperationResult, error) {
```

Comment:
After thinking about this, I think this function should be merged with `EnsureCreatedScenarioIsValid` since we're dealing with the same scenario and updating the same status which would be problematic to do in parallel - e.g. if one function decides that the scenario is fine, while the other one finds a problem and they both try to update the scenario status at nearly the same time.

Reply:
Merged with EnsureCreatedScenarioIsValid, dropped this ensure function at the same time.